### PR TITLE
impl: add universe domain options

### LINKED ISSUE #13116: define `UniverseDomainOption` in an internal namespace

### LINKED ISSUE #13117: define `UniverseDomainRetryPolicy` and `UniverseDomainRetryPolicyOption` in an internal namespace

### PR DESCRIPTION
**Pull request by scotthart**
fixes #13116 
fixes #13117

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120)

### CODE REVIEW

File: google/cloud/universe_domain_options.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+struct UniverseDomainOption {
+  using Type = std::string;
+};
+
+class UniverseDomainRetryPolicy : public ::google::cloud::RetryPolicy {
```
Comment (dbolduc) [inline google/cloud/universe_domain_options.h:32]: Will we offer defaults (`LimitedTime` / `LimitedErrorCount`) later?

File: google/cloud/universe_domain_options.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "google/cloud/version.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+struct UniverseDomainOption {
```
Comment (coryan) [inline google/cloud/universe_domain_options.h:28]: These should have Doxygen comments, as they will become public APIs at some point, right?

File: google/cloud/universe_domain_options.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+class UniverseDomainRetryPolicy : public ::google::cloud::RetryPolicy {
+ public:
+  virtual std::unique_ptr<UniverseDomainRetryPolicy> clone() const = 0;
+};
+
+struct UniverseDomainRetryPolicyOption {
+  using Type = std::shared_ptr<UniverseDomainRetryPolicy>;
+};
+
```
Comment (coryan) [inline google/cloud/universe_domain_options.h:40]: Most likely you will need a backoff option too?

**Review by dbolduc — APPROVED** [review]
Consider documenting the APIs now, even if we can't reference things like code snippets.

Approving so I don't block other PRs that need to use these interfaces.

File: google/cloud/universe_domain_options.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#include "google/cloud/retry_policy.h"
+#include "google/cloud/version.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
```
Comment (devbww) [inline google/cloud/universe_domain_options.h:26]: I assume this is temporary, and hence why this file isn't in `google/cloud/internal`.  Perhaps that deserves a temporary note too.

File: google/cloud/universe_domain_options.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+struct UniverseDomainOption {
+  using Type = std::string;
+};
+
+class UniverseDomainRetryPolicy : public ::google::cloud::RetryPolicy {
+ public:
+  virtual std::unique_ptr<UniverseDomainRetryPolicy> clone() const = 0;
+};
```
Comment (devbww) [inline google/cloud/universe_domain_options.h:35]: Why is this universe-domain specific.  It looks like a "CloneableRetryPolicy".

File: google/cloud/universe_domain_options.h (hunk 1 above)
Comment (devbww) [inline google/cloud/universe_domain_options.h:32]: `s/::google::cloud:://`?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @coryan, @dbolduc, and @devbww)

___
*[`google/cloud/universe_domain_options.h` line 26 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391449238:-Nj8mMDw2k8fMvHOf4_R:b-t777pj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L26)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I assume this is temporary, and hence why this file isn't in `google/cloud/internal`.  Perhaps that deserves a temporary note too.
</blockquote></details>

TODO with issue number added.
___
*[`google/cloud/universe_domain_options.h` line 28 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391433831:-Nj8mOcmAcxhco0yxcFr:bqmjwcz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L28)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

These should have Doxygen comments, as they will become public APIs at some point, right?
</blockquote></details>

Added.
___
*[`google/cloud/universe_domain_options.h` line 32 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391432946:-Nj8mPhw38naq-LaoHJf:bkvnygz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L32)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Will we offer defaults (`LimitedTime` / `LimitedErrorCount`) later?
</blockquote></details>

Maybe? There will be a default policy for when the user does not specify one.
___
*[`google/cloud/universe_domain_options.h` line 32 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391451140:-Nj8mXVmAU63Ju-t73pG:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L32)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

`s/::google::cloud:://`?
</blockquote></details>

Done.
___
*[`google/cloud/universe_domain_options.h` line 35 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391450461:-Nj8mZEn8eEaXlgEtNHK:b-y1mrmo) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L35)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Why is this universe-domain specific.  It looks like a "CloneableRetryPolicy".
</blockquote></details>

As far as I can tell, this is how we typically define a cloneable retry policy.
___
*[`google/cloud/universe_domain_options.h` line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391434133:-Nj8mYRxE-hK4YH39huF:bqmjwcz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Most likely you will need a backoff option too?
</blockquote></details>

Added.

File: google/cloud/universe_domain_options.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+// TODO(#13115): Remove internal namespace.
+namespace internal {
+
+/**
+ * Use with `google::cloud::Options` to configure the universe domain used in
```
Comment (dbolduc) [inline google/cloud/universe_domain_options.h:31]: Comment: argh. Links would be nice, but I doubt any public docs exist.

**Review by coryan — APPROVED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120)* status: all files reviewed, 5 unresolved discussions (waiting on @dbolduc, @devbww, and @scotthart)

___
*[`google/cloud/universe_domain_options.h` line 35 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13120#gh-1391450461:-Nj8vBrH3N0NfFJgB3uX:br51eu8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/746986ec87e59ccdb384aa279525a1d7fe53f71d/google/cloud/universe_domain_options.h#L35)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

As far as I can tell, this is how we typically define a cloneable retry policy.
</blockquote></details>

Our retry policies evolved such that "each library has its own hierarchy of retry policies".  We may do this differently if we were starting over. And we may change it too. Methinks this is not the PR to make that change.

**Pull request merged**